linked_list/linked_list.py

```diff
@@ -9,71 +9,148 @@ def __init__(self, value, next_node = None):
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def get_first(self):
-        pass
+        if not self.head:
+            return None
+        
+        return self.head.value
 
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(self, value):
-        pass
+        new_node = Node(value)
+        new_node.next = self.head
+        self.head = new_node
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(self, value):
-        pass
+        if not self.get_first():
+            return False
+        
+        current_node = self.head
+
+        while current_node != None:
+            if current_node.value == value:
+                return True
+            else:
+                current_node = current_node.next
+        
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def length(self):
-        pass
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
+    def length(self):        
+        counter = 0
+        current_node = self.head
+
+        while current_node != None:
+            counter += 1
+            current_node = current_node.next
+        
+        return counter
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        if self.length() < index:
+            return None
+        
+        current_node = self.head
+        count = 0
+        while count < index:
+            current_node = current_node.next
+            count += 1
+
+        return current_node.value
+
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        current_node = self.head
+
+        if current_node == None:
+            return None
+
+        while current_node.next != None:
+            current_node = current_node.next
+        
+        return current_node.value
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def add_last(self, value):

```

Comment:
✨ In keeping with the choices you made in other methods, I would return `None`in this method or just not explicitly (which will return None by default) with your rather than returning `self` which is the entire LinkedList object.